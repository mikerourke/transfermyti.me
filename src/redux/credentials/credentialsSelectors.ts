import { toolNameByMappingSelector } from "~/redux/allEntities/allEntitiesSelectors";
import { createSelector } from "~/redux/reduxTools";
import {
  FetchStatus,
  Mapping,
  ToolName,
  type Credentials,
  type CredentialsByMapping,
  type ReduxState,
  type ValidationErrorsByMapping,
} from "~/types";

export const validationFetchStatusSelector = createSelector(
  (state: ReduxState) => state.credentials.validationFetchStatus,
  (validationFetchStatus): FetchStatus => validationFetchStatus,
);

export const validationErrorsByMappingSelector = createSelector(
  (state: ReduxState) => state.credentials.validationErrorsByMapping,
  (validationErrorsByMapping): ValidationErrorsByMapping =>
    validationErrorsByMapping,
);

export const hasValidationErrorsSelector = createSelector(
  validationErrorsByMappingSelector,
  (validationErrorsByMapping): boolean =>
    Object.values(validationErrorsByMapping).some(Boolean),
);

const sourceCredentialsSelector = createSelector(
  (state: ReduxState) => state.credentials.source,
  (sourceCredentials) => sourceCredentials,
);

const targetCredentialsSelector = createSelector(
  (state: ReduxState) => state.credentials.target,
  (targetCredentials) => targetCredentials,
);

export const credentialsByMappingSelector = createSelector(
  sourceCredentialsSelector,
  targetCredentialsSelector,
  (sourceCredentials, targetCredentials): CredentialsByMapping => ({
    [Mapping.Source]: sourceCredentials,
    [Mapping.Target]: targetCredentials,
  }),
);

export const credentialsByToolNameSelector = createSelector(
  toolNameByMappingSelector,
  sourceCredentialsSelector,
  targetCredentialsSelector,
  (
    toolNameByMapping,
    sourceCredentials,
    targetCredentials,
  ): Record<ToolName, Credentials> =>
    ({
      [toolNameByMapping[Mapping.Source]]: sourceCredentials,
      [toolNameByMapping[Mapping.Target]]: targetCredentials,
    } as Record<ToolName, Credentials>),
);
