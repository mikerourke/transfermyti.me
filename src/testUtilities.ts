export { FAKES } from "~/redux/__fakes__/state";

type MockInstance = jest.MockInstance<AnyValid, AnyValid>;

interface WrapForMockingReturn {
  calls: AnyValid[];
  clear(): MockInstance;
  implementOnce(funcOrReturn: AnyValid): MockInstance;
  rejectAlways(err?: AnyValid): MockInstance;
  rejectOnce(err?: AnyValid): MockInstance;
  resolveMultiple(values: AnyValid[]): void;
  resolveOnce(value?: AnyValid): MockInstance;
  returnOnce(value?: AnyValid): MockInstance;
  returnMultiple(values: AnyValid[]): void;
  throwOnce(message?: string): MockInstance;
}

/**
 * Wrappers around someFunc.mock...() that Jest provides. Using
 * this avoids the annoying `someFunc as AnyValid` required to shut TypeScript
 * up.
 */
export function wrapForMocking(func: AnyValid): WrapForMockingReturn {
  return {
    calls: func?.mock?.calls,
    clear(): MockInstance {
      return func.mockClear();
    },
    implementOnce(funcOrReturn: AnyValid): MockInstance {
      if (typeof funcOrReturn === "function") {
        return func.mockImplementationOnce(funcOrReturn);
      } else {
        return func.mockImplementationOnce(() => funcOrReturn);
      }
    },
    rejectAlways(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValue(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    rejectOnce(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValueOnce(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    resolveMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockResolvedValueOnce(value);
      }
    },
    resolveOnce(value: AnyValid = null): MockInstance {
      return func.mockResolvedValueOnce(value);
    },
    returnOnce(value: AnyValid = null): MockInstance {
      return func.mockReturnValueOnce(value);
    },
    returnMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockReturnValueOnce(value);
      }
    },
    throwOnce(message: string = "error"): MockInstance {
      return func.mockImplementationOnce(() => {
        throw new Error(message);
      });
    },
  };
}

/**
 * Allows you to get around `delay()` calls in sagas.
 * @see https://github.com/jfairbank/redux-saga-test-plan/issues/257#issuecomment-503016758
 * @example
 *   return expectSaga(someSaga)
 *     .provide([
 *       { call: provideDelay },
 *     ])
 *     ...
 */
export function provideDelay(
  { fn }: AnyValid,
  next: AnyValid,
): null | VoidFunction {
  return fn.name === "delayP" ? null : next();
}
