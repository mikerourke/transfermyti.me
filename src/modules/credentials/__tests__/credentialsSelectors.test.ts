import cases from "jest-in-case";

import { FAKES } from "~/testUtilities";

import * as credentialsSelectors from "../credentialsSelectors";

const { REDUX_STATE } = FAKES;

describe("within credentialsSelectors", () => {
  cases(
    "selectors that directly access state return the correct value",
    (options) => {
      const result = options.selector(REDUX_STATE);

      expect(result).toEqual(options.expected);
    },
    [
      {
        name: "validationFetchStatusSelector returns state.validationFetchStatus",
        selector: credentialsSelectors.validationFetchStatusSelector,
        expected: REDUX_STATE.credentials.validationFetchStatus,
      },
      {
        name: "validationErrorsByMappingSelector returns state.validationErrorsByMapping",
        selector: credentialsSelectors.validationErrorsByMappingSelector,
        expected: REDUX_STATE.credentials.validationErrorsByMapping,
      },
    ],
  );

  test("the hasValidationErrorsSelector returns true if validation errors present in state", () => {
    const result = credentialsSelectors.hasValidationErrorsSelector({ ...REDUX_STATE });

    expect(result).toBe(true);
  });

  cases(
    "the selectors match their snapshots",
    (options) => {
      const result = options.selector({ ...REDUX_STATE });

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the credentialsByMappingSelector",
        selector: credentialsSelectors.credentialsByMappingSelector,
      },
      {
        name: "for the credentialsByToolNameSelector",
        selector: credentialsSelectors.credentialsByToolNameSelector,
      },
    ],
  );
});
