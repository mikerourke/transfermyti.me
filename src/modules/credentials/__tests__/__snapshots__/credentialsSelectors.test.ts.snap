// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within credentialsSelectors the selectors match their snapshots for the credentialsByMappingSelector 1`] = `
{
  "source": {
    "apiKey": "TOGGLAPIKEY",
    "email": "test-user@test.com",
    "userId": "6001",
  },
  "target": {
    "apiKey": "CLOCKIFYAPIKEY",
    "email": "test-user@test.com",
    "userId": "clock-user-01",
  },
}
`;

exports[`within credentialsSelectors the selectors match their snapshots for the credentialsByToolNameSelector 1`] = `
{
  "clockify": {
    "apiKey": "CLOCKIFYAPIKEY",
    "email": "test-user@test.com",
    "userId": "clock-user-01",
  },
  "toggl": {
    "apiKey": "TOGGLAPIKEY",
    "email": "test-user@test.com",
    "userId": "6001",
  },
}
`;
